Release notes for 0.21.0 (#32)
pyodide/pyodide-blog: content/posts/0.21-release.md

comments: false
# description: ""
disableHLJS: true # to disable highlightjs
disableShare: false
hideSummary: false
searchHidden: true
ShowReadingTime: true
ShowBreadCrumbs: true
ShowPostNavLinks: true
---

Pyodide 0.21 is a major release focused on improvements to the systems for
building and loading packages. We added support for out of tree package builds
which should soon make it much easier for people to build binary packages that
we do not include in our distribution. We added support for Rust packages and
made major progress in supporting CMake packages. We also added 34 new packages
including a large number of popular packages, for example bitarray,
opencv-python, shapely, and xgboost.

We moved from Emscripten v2.0.27 to Emscripten v3.1.14 which includes almost all
of the patches we need (we still apply two patches). There were also many minor
improvements and bugfixes to the Pyodide module and type conversions.

For the complete list of changes see the
[changelog](https://pyodide.org/en/stable/project/changelog.html#version-0-21-0).

## Out of tree binary package builds for Pyodide

The most exciting new feature in Pyodide 0.21.0 is that it adds experimental

> support for building binary packages for Pyodide "out of tree".

Reviewer: ```suggestion
support for building binary packages as wheels for Pyodide "out of tree", similarly to how standard Python wheels are built.
```

We need to emphasize that we are just building wheels, with a workflow close to the usual. "out of tree" doesn't really mean much for an average Python user.

BTW do we have docs for it? https://pyodide.org/en/stable/development/new-packages.html makes no mention of it. Maybe we should make this feature more prominent.

Pyodide has so far used a monolothic collection of package recipes inspired by
Conda to build binary extensions. Using a collection of package recipes has
significant advantages, allowing us to maintain patches and test all of the
packages together. However, it also has several downsides:

1. It requires all packages to be maintained in a monorepo.

> 2. It doesn't allow packages to support Pyodide in their CI.

Reviewer: ```suggestion
2. It doesn't easily allow packages to support Pyodide in their CI.
```
I guess it's not impossible, just not very practical.

> 3. There is no convenient way for someone to use a personal-use binary extension
>    with Pyodide.

Reviewer: ```suggestion
3. There is no convenient way for someone to use a private package with a  binary extension
   with Pyodide.
```

4. There is no good way to have several different versions of the same package.
   For instance, this would be very helpful for using Pyodide to make
   interactive documentation for old versions of a package.

We can build most packages out of tree, though this feature is still a bit
experimental. We will continue using the collection of package recipes for our

> in-tree builds. Our goal is to add Pyodide to the CI of key packages, for

Reviewer: Maybe it would be good to add an example here to make it more concrete. And also emphasize that this no longer requires downloading the pyodide repo.

For instance, here is a minimal setup to build a package wheel in Github Actions CI,
```yaml
runs-on: ubuntu-latest
    steps:
      -  uses: actions/checkout@v3
      - uses: actions/setup-python@v4
        with:
          python-version: 3.10.2
      - uses: mymindstorm/setup-emsdk@v11
        with:
          version: 3.1.14
      - run: pip install pyodide-build==0.21.0
      - run: pyodide build
```

Opened https://github.com/pyodide/pyodide/issues/2949

instance as a proof of concept we have added Pyodide to the Numpy CI in a draft
PR. In order to limit CI time, we only run a small fragment of each package's
tests in Pyodide's CI. However, the package's CI can run all of its tests in
Pyodide. This should help tremendously with locating and fixing subtle
Pyodide-only bugs.

## Improvements to package loading

> In order to support loading of packages built out of tree, we updated micropip
> to be able to load binary wheels and to be able to generate custom lock files.

Reviewer: ```suggestion
We have also updated micropip
to be able to load binary wheels (which must have a filename ending in `-cp310-cp310-emscripten_wasm32.whl`) and to be able to generate custom lock files.
```

We should probably add a section about generating lock files to https://pyodide.org/en/latest/usage/loading-packages.html#micropip

Opened https://github.com/pyodide/pyodide/issues/2948

We also added the capability of specifying custom lock files in `loadPyodide`.

> Custom lock files feature may help significantly with loading time for certain

Reviewer: Does it help much with load time though? As far as I understand it should be equivalent to running micropip.install anyway no? https://github.com/pyodide/pyodide/issues/2948

Author: Answered there.

applications and can also be useful for people who want to add extra packages to
their repls to be loaded with `loadPackagesFromImports`.

## Rust and CMake support

> Pyodide can now build Rust packages! Not only that, but PyO3 packages just work

Reviewer: Maybe add a link to PyO3, people don't necessarily know what it is.
